trl/extras/best_of_n_sampler.py

```diff
@@ -0,0 +1,76 @@
+from typing import Any, Dict, List, Optional, Union
+
+import torch
+from transformers import GenerationConfig, Pipeline, PreTrainedTokenizer, PreTrainedTokenizerFast
+
+from ..core import set_seed
+from ..models import SUPPORTED_ARCHITECTURES, PreTrainedModelWrapper
+
+
+class BestOfNSampler(object):
+    def __init__(
+        self,
+        model: PreTrainedModelWrapper,
+        tokenizer: Union[PreTrainedTokenizer, PreTrainedTokenizerFast],
+        reward_pipeline: Pipeline,
+        sample_size: int = 4,
+        length_sampler: Any = None,
+        seed: Optional[int] = None,
+        generation_config: Optional[GenerationConfig] = None,
+        reward_kwargs: Dict[str, Any] = {},
+    ) -> None:
+        if seed is not None:
+            set_seed(seed)
+
+        if not isinstance(tokenizer, (PreTrainedTokenizer, PreTrainedTokenizerFast)):
+            raise ValueError(
+                f"tokenizer must be a PreTrainedTokenizer or PreTrainedTokenizerFast, got {type(tokenizer)}"
+            )
+        if not isinstance(model, (SUPPORTED_ARCHITECTURES)):
+            raise ValueError(
+                f"model must be a PreTrainedModelWrapper, got {type(model)} - supported architectures are: {SUPPORTED_ARCHITECTURES}"
+            )
+
+        self.model = model
+        self.tokenizer = tokenizer
+
+        if reward_pipeline is None:
+            raise ValueError("reward_pipeline must be provided")
+
```

Comment:
After thinking a bit, can we just make the reward a callable instead of using pipeline?

Reply:
I wasn't aware there would be such a use-case, thank you for that and will do

Reply:
I think the reasoning to make more broader makes sense. 

Just to clarify, and I may be skipping a few steps and with the constraint of making the API look less like a patchwork doll, do you envision removing all the pipeline based parameters :`reward_pipeline` and  `reward_kwargs` and with this also removes any assumption we can make about the pipeline output in this line (and we remove this as well) 

`[self.reward_extractor(output) for output in self.reward_pipeline(output, **self.reward_kwargs)]` ?

This would result in placing the onus on the user to give us a callable that takes care of all the three concerns (model, model params and output extraction) so all we have to do is plug it here: https://github.com/lvwerra/trl/pull/375/files#diff-667e046b6a4c1712553bdec501359fe107cb415f05e5b2526b1df83fa2b4d14bR72